theano/gpuarray/dnn.py

```diff
@@ -306,15 +309,46 @@ def __init__(self, border_mode, subsample=(1, 1), conv_mode='conv',
         self.subsample = subsample
         assert conv_mode in ('conv', 'cross')
         self.conv_mode = conv_mode
+        if conv_mode == 'conv':
+            conv_mode = 0  # CUDNN_CONVOLUTION
+        else:
+            conv_mode = 1  # CUDNN_CROSS_CORRELATION
 
-        assert precision in ['float16', 'float32', 'float64']
-        self.precision = precision
-
-    def make_node(self, kern_shape):
-        if kern_shape.type.ndim != 1 or kern_shape.type.dtype != 'int64':
-            raise TypeError('kern must be 1D shape tensor')
-
-        node = Apply(self, [kern_shape],
+        if precision not in ['float16', 'float32', 'float64']:
+            raise TypeError('precision must be one of float16, float32, float64')
+        else:
+            self.precision = precision
+            if precision == 'float16':
+                precision = 16
+            elif precision == 'float32':
+                precision = 32
+            elif precision == 'float64':
+                precision = 64
+        if isinstance(border_mode, tuple) or border_mode == 'valid':
+            bmode = 1
+        elif border_mode == 'half':
+            bmode = 2
+        elif border_mode == 'full':
+            bmode = 0
+        else:
+            raise ValueError("Invalid value for border_mode")
+        padding = [0, 0, 0]
+        if isinstance(border_mode, tuple):
+            padding[0] = border_mode[0]
+            padding[1] = border_mode[1]
+            if len(border_mode) > 2:
+                padding[2] = border_mode[2]
+        else:
+            padding = [0, 0, 0]
+        nb_dims = as_tensor_variable(len(subsample))
+        if len(subsample) == 2:
+            subsample += (0,)
+        padding = as_tensor_variable(padding)
+        subsample = as_tensor_variable(list(subsample))
+        conv_mode = as_tensor_variable(conv_mode)
+        precision = as_tensor_variable(precision)
+        bmode = as_tensor_variable(bmode)
```

Comment:
You must make sure in the make_node that they are of the right dtype. Can you do:
subsample = as_tensor_variable(...).astype('int64')

to be sure they are of the right dtype. Otherwise, your cast in the c code to int64 could give the wrong answer in some cases.